CDH: Implement I2C and tinyprotocol basic API

BrendanKelly84213 opened the pull request:
## Description (tell us what you changes or added or removed please)
NOTE: This now hard depends on #82 ! Please no merge before merging that!!

This is basically a wrapper around tinyprotocol. Exposes the functionality we need and performs initialization.

## How to test (if any tests done)
Here's the example main file I was using for testing

```C
#include "main.h"

#include <hal_delay.h>
#include <CDH/driver_init.h>
#include <CDH/pins.h>
#include <commsdrive.h>

#define SLAVE_ADDR 0x08

int main(void)
{
        /* Initializes MCU, drivers and middleware */
        system_init();
        commsdrive_init();

        /* Replace with your application code */
        static const uint8_t size = 5;
        uint8_t rx_buf[size];
        volatile int32_t ret = 0;
        while (1) {
                gpio_toggle_pin_level(LED0);
                ret = commsdrive_process_telerequest(SLAVE_ADDR, 0x01, rx_buf, size);
                delay_ms(100);
        }
}

```

This just gets the system status and firmware version from BMS, since those should be static values and are easy to check. BMS has to be current for this to work, in other words https://github.com/scsd-cdh/OBC/pull/77 has to have been merged. You can look at the values with a debugger or use the logic analyzer, whatever you want.



## Checklist
- [x] Code Compiles
- [x] Did you add any comments
- [x] Did you test it at all ?! Good Boy/Girl

dqwed01 commented on MIDDLEWARE/commsdrive/commsdrive.c:
```diff
@@ -0,0 +1,48 @@
+	TINYPROTOCOL_Initialize();
+}
+
+void CommsDrive_ProcessTeleRequest(uint8_t slaveAddr, uint8_t request, uint8_t* rx_buf, uint8_t len)
```
Would be nice to have a function header just quickly explain what this is doing and why. Same thing for any additional complex functions moving forward

BrendanKelly84213 replied:
Wait you want a description of how this works? Added a comment in the header describing what it does. Is something not clear in the implementation? Let me know what if so I can add a little inline explenation

dqwed01 replied:
Probably an inline approach would be better, explaining the the purpose of the code and what kind of output you'd expect from it. Doesn't have to be super lengthy, just a small description so we're not completely lost trying to understand the code

BrendanKelly84213 replied:
Anything specific?

silvehh commented on MIDDLEWARE/commsdrive/commsdrive.c:
```diff
@@ -0,0 +1,75 @@
+	Mas_I2C_0_PORT_init();
+}
+
+static int16_t transmitI2C(const uint8_t* buffer, uint8_t size)
```
I see trasmitI2C but where is receiveI2C?

BrendanKelly84213 replied:
This is just the callback tinyprotocol requires to do I2C write. tinyprotocol doesn't support i2c read as far as I can tell, we do it manually in CommsDrive_ProcessTeleRequest, after we do the write. Like we i2c write to tell the slave what we want, then we i2c read to get the data

silvehh reviewed (commented):
I see I2C wrapper is intertwined with commsdrive. This may not be good structure as it is not modular, I suppose SPI, LVDS, etc will also be added in there? It would be better to have an I2Cwrapper.c, SPIwrapper.c, etc as to be most modular and reusable. Make sure code is reusable; with the current architecture, commsdrive seems to be non-generic. For example, lets say I want to use SPI only for whatever use case, why should I have to called commsdrive_init, which initialises all comms in comm drive. I suggest separating all comms protocols and having separate inits, this will incentivise reusability in future.

BrendanKelly84213 committed:
CDH: I2C Wrapper

Adds a wrapper around I2C functionality for CDH samv71 board. Support
master for now.

CDH: WIP consistent i2c write

BrendanKelly84213 committed:
CDH: I2C commnunication commsdrive module

Basically a wrapper around tinyprotocol. Lets us do telemetry requests
telecommands and handles sending and receiving and packaging,
initializes i2c and tinyprotocol, etc...

WIP

BrendanKelly84213 committed:
TinyProtocol: CMake support

BrendanKelly84213 committed:
TinyProtocol: Support master writing to specific slave

dqwed01 reviewed (approved):
lgtm
